polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -767,6 +856,142 @@ class RaiseSetKernel : public OpInterfaceHostRaisePattern<FunctionOpInterface> {
+      if (!(annotationStr == "nd_range" || annotationStr == "offset" ||
+            annotationStr == "range"))
+        return WalkResult::advance();
+
+      // Get value's constructor
+      Value value = annotation.getVal();
+      sycl::SYCLHostConstructorOp constructor = getConstructor(value);
+      if (!constructor)
+        return WalkResult::interrupt();
+
+      assert((TypeSwitch<Type, bool>(constructor.getType().getValue())
+                  .Case<sycl::NdRangeType>(
+                      [=](auto) { return annotationStr == "nd_range"; })
+                  .Case<sycl::IDType>(
+                      [=](auto) { return annotationStr == "offset"; })
+                  .Case<sycl::RangeType>(
+                      [=](auto) { return annotationStr == "range"; })
+                  .Default(false)) &&
+             "Invalid constructor type");
+
+      constructors.push_back(constructor);
+      annotations.push_back(annotation);
+
+      return WalkResult::advance();
+    };
+    if (op.walk(getConstructors).wasInterrupted() || constructors.empty())
+      return failure();
+
+    // We will introduce the operation after the last annotation
+    OpBuilder::InsertionGuard ig(rewriter);
+    rewriter.setInsertionPointAfter(annotations.back());
+    Location loc = annotations.back().getLoc();
+
+    // Remove annotations, no longer needed
+    // Also prevent infinite recursion
+    rewriter.updateRootInPlace(op, [&] {
+      for (LLVM::VarAnnotation annotation : annotations)
+        rewriter.eraseOp(annotation);
+    });
+
+    assert((constructors.size() < 3 &&
+            isa<sycl::NdRangeType, sycl::RangeType>(
+                constructors.front().getType().getValue()) &&
+            (constructors.size() == 1 ||
+             (isa<sycl::RangeType>(constructors.front().getType().getValue()) &&
+              isa<sycl::IDType>(constructors.back().getType().getValue())))) &&
+           "Expecting nd-range or global size+[offset] combination");
+
+    // TODO: nd_range constructor receives pointers to structs, so we are not
+    // able to handle this yet.
+    if (isa<sycl::NdRangeType>(constructors.front().getType().getValue()))
+      return failure();
+
+    SmallVector<Value> arguments = getArguments(op, constructors, rewriter);
+
+    // Finally insert sycl.host.handler.set_nd_range after the last annotation
+    rewriter.updateRootInPlace(op, [&] {
+      rewriter.create<sycl::SYCLHostHandlerSetNDRange>(loc, handler, arguments);
+    });
+
+    return success();
+  }
+
+private:
+  static SmallVector<Value>
+  getArguments(FunctionOpInterface op,
+               ArrayRef<sycl::SYCLHostConstructorOp> constructors,
+               PatternRewriter &rewriter) {
+    SmallVector<Value> arguments;
+    OpBuilder::InsertionGuard ig(rewriter);
+    llvm::transform(
+        constructors, std::back_inserter(arguments),
+        [&](sycl::SYCLHostConstructorOp constructor) -> Value {
```

Comment:
Not sure what happened to my previous comment, but I still think it we could extract this lambda to a function to aid readability.

Reply:
I extracted the bigger lambda to another function. Extracting the lambda being passed to transform would require creating a struct. Not sure that's more readable to what we have now (we're capturing `rewriter`.